substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -1704,6 +1815,381 @@ impl<T: Config> Pallet<T> {
+        // then contract is canceled and removed from service contract map
+        Self::_service_contract_cancel(
+            account_id,
+            service_contract_id,
+            types::Cause::CanceledByUser,
+        )?;
+
+        Ok(().into())
+    }
+
+    pub fn _service_contract_cancel(
+        account_id: T::AccountId,
+        service_contract_id: u64,
+        cause: types::Cause,
+    ) -> DispatchResultWithPostInfo {
+        let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
+            .ok_or(Error::<T>::TwinNotExists)?;
+
+        let service_contract = ServiceContracts::<T>::get(service_contract_id)
+            .ok_or(Error::<T>::ServiceContractNotExists)?;
+
+        // Only service or consumer can cancel contract
+        ensure!(
+            twin_id == service_contract.service_twin_id
+                || twin_id == service_contract.consumer_twin_id,
+            Error::<T>::TwinNotAuthorized,
+        );
+
+        // Remove contract from service contract map
+        // Can be done at any state of contract
+        // so no need to handle state validation
+        ServiceContracts::<T>::remove(service_contract_id);
+        Self::deposit_event(Event::ServiceContractCanceled {
+            service_contract_id,
+            cause,
+        });
+
+        log::info!(
+            "successfully removed service contract with id {:?}",
+            service_contract_id,
+        );
+
+        Ok(().into())
+    }
+
+    #[transactional]
+    pub fn _service_contract_bill(
+        account_id: T::AccountId,
+        service_contract_id: u64,
+        variable_amount: u64,
+        metadata: Vec<u8>,
+    ) -> DispatchResultWithPostInfo {
+        let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
+            .ok_or(Error::<T>::TwinNotExists)?;
+
+        let mut service_contract = ServiceContracts::<T>::get(service_contract_id)
+            .ok_or(Error::<T>::ServiceContractNotExists)?;
+
+        // Only service can bill consumer for service contract
+        ensure!(
+            twin_id == service_contract.service_twin_id,
+            Error::<T>::TwinNotAuthorized,
+        );
+
+        // Allow to bill contract only if approved by both
+        ensure!(
+            matches!(
+                service_contract.state,
+                types::ServiceContractState::ApprovedByBoth
+            ),
+            Error::<T>::ServiceContractBillingNotAllowed,
+        );
+
+        // Get elapsed time (in seconds) to bill for service
+        let now = <timestamp::Pallet<T>>::get().saturated_into::<u64>() / 1000;
+        let elapsed_seconds_since_last_bill = now - service_contract.last_bill;
+
+        // Billing time (window) is max 1h by design
+        // So extra time will not be billed
+        // It is the service responsability to bill on right frequency
+        let window = elapsed_seconds_since_last_bill.min(3600);
```

Comment:
Or even make it a config parameter type?

Reply:
> Or even make it a config parameter type?

I did the first option for now
Not sure if I see what you mean by the second option

Reply:
It's a parameter that you can set in the definition of the config of a pallet. Like here for example: https://github.com/threefoldtech/tfchain/blob/development/substrate-node/runtime/src/lib.rs#L381

Reply:
I added such config parameter type here
https://github.com/threefoldtech/tfchain/pull/522/commits/70bfb6007789076d5e4287acfcb919d7a94ec22a
@brandonpille @DylanVerstraete please check if every thing is ok because changes can be critical ;)

Reply:
Looks great. Although maybe ServiceContractBillingFrequency would be a better name?

Reply:
> maybe ServiceContractBillingFrequency would be a better name?

Not sure because it is used in other contexts
The idea is to characterize the amount of time (is seconds) price values are set for in pricing policy / billing context 
maybe `PriceReferencePeriod` would be better?

Reply:
Oh I didn't know that. Then the name is fine how it is.